source/socks5-support/socks5.rst

```diff
@@ -0,0 +1,192 @@
+otherwise be unreachable from the local network by connecting to a proxy
+server, which receives the intended target host’s address from the client
+and then connects to that address.
+
+This specification defines driver behaviour when connecting to MongoDB services
+through a SOCKS5 proxy.
+
+META
+====
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in `RFC 2119 <https://www.ietf.org/rfc/rfc2119.txt>`__.
+
+Specification
+=============
+
+
+Terms
+-----
+
+SOCKS5
+^^^^^^
+
+The SOCKS protocol, version 5, as defined in `RFC1928 <https://datatracker.ietf.org/doc/html/rfc1928>`__,
+restricted to either no authentication or username/password authentication
+as defined in `RFC1929 <https://datatracker.ietf.org/doc/html/rfc1929>`__.
+
+
+MongoClient Configuration
+-------------------------
+
+proxyHost
+^^^^^^^^^
+
+To specify to the driver to connect using a SOCKS5 proxy, a connection string
+option of :code:`proxyHost=host` MUST be added to the connection string
+or passed through an equivalent :code:`MongoClient` option.
+This option specifies a domain name or IPv4 or IPv6 address on which
+a SOCKS5 proxy is listening.
+This option MUST only be configurable at the level of a :code:`MongoClient`.
+
+proxyPort
+^^^^^^^^^
+
+To specify to the driver to connect using a SOCKS5 proxy listening
+on a non-default port, a connection string option of :code:`proxyPort=port`
+MUST be added to the connection string or passed through an
+equivalent :code:`MongoClient` option.
+This option specifies a TCP port number. The default for this option
+MUST be :code:`1080`.
+This option MUST only be configurable at the level of a :code:`MongoClient`.
+Drivers MUST error if this option was specified and :code:`proxyHost`
+was not specified.
+
+proxyUsername
+^^^^^^^^^^^^^
+
+To specify to the driver to connect using a SOCKS5 proxy requiring
+username/password authentication, a connection string option of
+:code:`proxyUsername=username` MUST be added to the connection string
+or passed through an equivalent :code:`MongoClient` option.
+This option specifies a string of non-zero length. Drivers MUST ignore
+this option if it specifies a zero-length string. Drivers MUST error
+if this option was specified and :code:`proxyHost` was not specified
+or :code:`proxyPassword` was not specified.
+
+proxyPassword
+^^^^^^^^^^^^^
+
+To specify to the driver to connect using a SOCKS5 proxy requiring
+username/password authentication, a connection string option of
+:code:`proxyPassword=password` MUST be added to the connection string
+or passed through an equivalent :code:`MongoClient` option.
+This option specifies a string of non-zero length. Drivers MUST ignore
+this option if it specifies a zero-length string. Drivers MUST error
+if this option was specified and :code:`proxyHost` was not specified
+or :code:`proxyUsername` was not specified.
```

Comment:
> Did we consider alternate URI options such as a single proxy option using a socks5 scheme?

> Or modelling it after authMechanismProperties?

I don’t think it was brought up during the scope review, no. I can see the advantages of both variants and I’m definitely open to it, though.

Is this something that could be discussed synchronously in some drivers meeting? It might be good to have a decision on this as soon as we can.

Reply:
I'd suggest a quick discussion / poll in #driver-devs to get a wider perspective - especially with respect to potential challenges for various URL parser implementations across drivers.

Reply:
Have we considered the security implications of putting a password inside URI options? Doing so may make it easy to leak the password in logs and error messages. Is there an alternative?

Reply:
Valid concern with respect to security implications of encoding complexities leading to a possible proxy password leak. Another idea would be to combine proxy host and port into a single option:

`?proxy=proxy1:8888&proxyUser=user1&proxyPassword=pwd1`

Reply:
> Have we considered the security implications of putting a password inside URI options? Doing so may make it easy to leak the password in logs and error messages. Is there an alternative?
> 
> Edit: I suppose we already have the same issue with the tlsCertificateKeyFilePassword option.

That and the `AWS_SESSION_TOKEN` as part of `authMechanismProperties` as well – we have logic in our developer tools to redact those strings specifically, and would add one for the proxy password as well, so yeah, it’s an existing problem.

> I'd suggest a quick discussion / poll in #driver-devs to get a wider perspective - especially with respect to potential challenges for various URL parser implementations across drivers.

I don’t have access to that channel, I guess #drivers is a bit of an odd place but it should work? https://mongodb.slack.com/archives/C0UD1CGUF/p1639392502073400

Reply:
Looks like separate parameters is the far and away winner unless poll numbers shift substantially. I'll LGTM but feel free to re-request my review if poll numbers swing radically and you decide to go with a different proxy options format.